
---

在 Debug 和 ReleaseSafe 模式中，`unreachable` 将会触发一个带有 `reached unreachable code` 信息的 `panic`。

在 ReleaseFast 和 ReleaseSmall 模式中，优化器会基于 `unreachabel` 代码永远不会被触发的假设进行优化。

```zig file:test_unreachable.zig
// unreachable is used to assert that control flow will never reach a
// particular location:
test "basic math" {
  const x = 1;
  const y = 2;
  if (x + y != 3) {
    unreachable;
  }
}
```

实际上，`std.debug.assert` 就是这样实现的：

```zig file:test_assertion_failure.zig
// This is how std.debug.assert is implemented
fn assert(ok: bool) void {
  if (!ok) unreachable; // assertion failure
}

// This test will fail because we hit unreachable.
test "this will fail" {
  asser(false);
}
```

```bash
$ zig test test_assertion_failure.zig
1/1 test_assertion_failure.test.this will fail... thread 2828429 panic: reached unreachable code
test_assertion_failure.zig:3:14: 0x103960d in assert (test)
    if (!ok) unreachable; // assertion failure
             ^
```

编译时的 `unreachable` 具有以下行为：

```zig file:test_comtime_unreachable.zig
const std = @import("std");
const assert = std.debug.assert;

test "type of unreachable" {
  comptime {
    // The type of unreachable is noreturn.
    // However this assertion will still fail to compile because
    // unreachable expressions are compile errors.
    assert(@TypeOf(unreachable) == noreturn);
  }
}
```

```bash
$ zig test test_comptime_unreachable.zig
test_comptime_unreachable.zig:10:16: error: unreachable code
        assert(@TypeOf(unreachable) == noreturn);
               ^~~~~~~~~~~~~~~~~~~~
test_comptime_unreachable.zig:10:24: note: control flow is diverted here
        assert(@TypeOf(unreachable) == noreturn);
                       ^~~~~~~~~~~
```
